colossalai/inference/tensor_parallel/engine.py

```diff
@@ -0,0 +1,250 @@
+
+    def __init__(self,
+                 model: nn.Module,
+                 max_batch_size: int,
+                 max_input_len: int,
+                 max_output_len: int,
+                 dtype: torch.dtype = torch.float16,
+                 device: str = 'cuda') -> None:
+        self.model = model
+        self.sharded_model = None
+
+        self.max_batch_size = max_batch_size
+        self.max_input_len = max_input_len
+        self.max_output_len = max_output_len
+        self.max_total_token_num = self.max_batch_size * (self.max_input_len + self.max_output_len)
+
+        # Constraints relatable with specs of devices
+        assert self.max_batch_size <= 64
+        assert self.max_input_len + self.max_output_len <= 2048
+
+        # NOTE For now, we focus on tensor parallel.
+        # We might want to merge pp and dp inference in future.
+        # self.tp_size = tp_size
+        # self.pp_size = 1
+        # self.dp_size = 1
+        torch.cuda.set_device(device=device)
+        self.dtype = dtype
+
+        self.head_dim = self.model.config.hidden_size // self.model.config.num_attention_heads
+        self.head_num = self.model.config.num_attention_heads
+        self.layer_num = self.model.config.num_hidden_layers
+
+        self.tp_size = -1    # toe be set with given shard config in self.prepare_shard_config
+
+    def _init_manager(self) -> None:
+        assert self.tp_size >= 1, "TP size not initialized without providing a valid ShardConfig"
+        assert self.head_num % self.tp_size == 0, f"Cannot shard {self.head_num} heads with tp size {self.tp_size}"
+        self.head_num //= self.tp_size    # update sharded number of heads
+        self.cache_manager = MemoryManager(self.max_total_token_num, self.dtype, self.head_num, self.head_dim,
+                                           self.layer_num)
+
+    def prepare_with_shard_config(self, shard_config: Optional[ShardConfig] = None) -> ShardConfig:
+        """ Prepare the engine with a given ShardConfig, or create a default one with tp size 1 """
+        if shard_config is None:
+            shard_config = ShardConfig(
+                tensor_parallel_process_group=None,
+                pipeline_stage_manager=None,
+                enable_tensor_parallelism=False,
+                enable_fused_normalization=False,
+                enable_all_optimization=False,
+                enable_flash_attention=False,
+                enable_jit_fused=False,
+                inference_only=True,
+            )
+            self.tp_size = 1
+        else:
+            shard_config.inference_only = True
+            shard_config.pipeline_stage_manager = None
+            if shard_config.enable_tensor_parallelism:
+                world_size = dist.get_world_size(shard_config.tensor_parallel_process_group)
+                self.tp_size = world_size
+        self._init_manager()
+
+        return shard_config
+
+    def shard_model_by(self, shardformer: ShardFormer) -> None:
```

Comment:
Since we don't want to set shard former inside the engine. Maybe we could consider the function, or a revised one, as an api to prompt for a shradformer and execute the sharding on the model the engine is holding

Reply:
sure